### PR TITLE
Authentication error messages

### PR DESCRIPTION
**Pull request by Fouad-sys**
This Pull Request displays specific error messages once the user tries to Login such as an incorrect email or password error if their password is incorrect or an invalid email if the email entered is not registered. It also displays Sign up error messages if an entered email is in an incorrect format or the password is too weak (less than 6 characters) or if a user with the typed email is already registered.

### CODE REVIEW

File: mobile/src/main/java/ch/epfl/sdp/mobile/infrastructure/persistence/auth/firebase/FirebaseAuth.kt [hunk 1]
```diff
@@ -38,28 +38,38 @@ class FirebaseAuth(private val actual: ActualFirebaseAuth) : Auth {
...
         val user = result?.user?.let(::FirebaseUser)
         AuthenticationResult.Success(user)
       } catch (other: Throwable) {
-        AuthenticationResult.FailureInternal
+        when (other) {
+          is FirebaseAuthWeakPasswordException -> AuthenticationResult.FailureBadPassword
+          is FirebaseAuthInvalidCredentialsException ->
+              when (type) {
+                "SignIn" -> AuthenticationResult.FailureIncorrectPassword
+                "SignUp" -> AuthenticationResult.FailureIncorrectEmailFormat
```
Comment (KurohanaJuri) [inline mobile/src/main/java/ch/epfl/sdp/mobile/infrastructure/persistence/auth/firebase/FirebaseAuth.kt:53]: I don't really like these magic values...(it's error prone) Maybe transform this into a enum

File: mobile/src/androidTest/java/ch/epfl/sdp/mobile/test/application/AuthenticationFacadeTest.kt [hunk 2]
```diff
@@ -43,7 +43,59 @@ class AuthenticationFacadeTest {
 
     val facade = AuthenticationFacade(auth, firestore)
 
-    assertThat(facade.signInWithEmail("email", "password")).isEqualTo(Failure)
+    assertThat(facade.signInWithEmail("email@epfl.ch", "password")).isEqualTo(FailureInvalidUser)
+  }
+
+  @Test
+  fun Given_nonEmptyAuth_When_theEnteredPasswordIsIncorrect_Then_aFailureIncorrectPasswordOccurs() {
```
Comment (KurohanaJuri) [inline mobile/src/androidTest/java/ch/epfl/sdp/mobile/test/application/AuthenticationFacadeTest.kt:50]: The name doesn't respect our convention. The keyword must be in lowercase.

File: mobile/src/main/java/ch/epfl/sdp/mobile/ui/i18n/English.kt [hunk 3]
```diff
@@ -14,6 +14,14 @@ object English : LocalizedStrings {
   override val authenticateSwitchToLogIn = "Log in".uppercase()
   override val authenticateSwitchToRegister = "Sign up".uppercase()
   override val authenticateErrorFailure = "Something went wrong when authenticating"
+  override val authenticationBadPasswordFailure =
+      "Your password should be at least 6 characters long, and contain at least one number, capital letter and special character"
```
Comment (KurohanaJuri) [inline mobile/src/main/java/ch/epfl/sdp/mobile/ui/i18n/English.kt:18]: Are you sure about that ? I see some valid password on some tests that don't have these requirements

Comment (Fouad-sys) [reply mobile/src/main/java/ch/epfl/sdp/mobile/ui/i18n/English.kt:18]: It actually doesn’t support the “Should contain at least one number, capital letter and special character as I realized the Firebase Error does not support that. But I can add that manually if we want to support it.

File: mobile/src/main/java/ch/epfl/sdp/mobile/application/authentication/AuthenticationResult.kt [hunk 4]
```diff
@@ -9,5 +9,22 @@ enum class AuthenticationResult {
...
+
+  /** Indicates that the given password associated to an email is incorrect. */
+  FailureIncorrectPassword,
+
+  /**
+   * Indicates that the user account corresponding to an email does not exist or has been disabled.
+   */
+  FailureInvalidUser,
+
   // TODO : Some specific failures could be added here, like FailureBadPassword, etc.
```
Comment (matt989253) [inline mobile/src/main/java/ch/epfl/sdp/mobile/application/authentication/AuthenticationResult.kt:29]: You can probably remove this TODO.
```suggestion
```

File: mobile/src/main/java/ch/epfl/sdp/mobile/infrastructure/persistence/auth/Auth.kt [hunk 5]
```diff
@@ -21,6 +21,24 @@ interface Auth {
...
+    /** Indicates that the given password associated to an email is incorrect. */
+    object FailureIncorrectPassword : AuthenticationResult
+
+    /**
+     * Indicates that the user account corresponding to an email does not exist or has been
+     * disabled.
+     */
+    object FailureInvalidUser : AuthenticationResult
+
     // TODO : Some specific failures could be added here, like FailureBadPassword, etc.
```
Comment (matt989253) [inline mobile/src/main/java/ch/epfl/sdp/mobile/infrastructure/persistence/auth/Auth.kt:42]: Here as well.
```suggestion
```

**Review by matt989253 — APPROVED** [review]
LGTM! Just a few `TODO`s that can be removed but that's it.

**Comment by matt989253** [comment]
Also, please don't forget to resolve conversations for which you've made the requested changes, it helps reviewers know their comments have been taken into consideration! For added points, you can reference the commit in question like this: 
"Fixed in [a362d0a](https://github.com/epfl-SDP/android/pull/219/commits/a362d0acb66c08acfda4a28c09a830109ead737d)" (You just need to copy-paste the beginning of the commit hash "a362d0a" for example)

File: mobile/src/main/java/ch/epfl/sdp/mobile/infrastructure/persistence/auth/Auth.kt [hunk 6]
```diff
@@ -21,6 +21,24 @@ interface Auth {
...
+    object FailureExistingAccount : AuthenticationResult
+
+    /** Indicates that the given password associated to an email is incorrect. */
+    object FailureIncorrectPassword : AuthenticationResult
+
+    /**
+     * Indicates that the user account corresponding to an email does not exist or has been
+     * disabled.
+     */
+    object FailureInvalidUser : AuthenticationResult
```
Comment (alexandrepiveteau) [inline mobile/src/main/java/ch/epfl/sdp/mobile/infrastructure/persistence/auth/Auth.kt:40]: Question : why not have a hierarchy that makes sure all failures are of the same type ? Something along the lines of :

```kotlin
sealed interface AuthenticationResult {
  data class Success(val user: User?) : AuthenticationResult
  sealed interface Failure : AuthenticationResult {
    object Internal : Failure
    object BadPassword : Failure
  }
}
```

File: mobile/src/main/java/ch/epfl/sdp/mobile/infrastructure/persistence/auth/AuthenticationMode.kt [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+
+/** An enumeration representing the [AuthenticationMode] that a user chose. */
+enum class AuthenticationMode {
+
+  /** The mode in which a user wants to login to their existing account. */
+  SignIn,
+
+  /** The mode in which a user wants to create a new account. */
+  SignUp,
+}
```
Comment (alexandrepiveteau) [inline mobile/src/main/java/ch/epfl/sdp/mobile/infrastructure/persistence/auth/AuthenticationMode.kt:11]: It looks like this `AuthenticationMode` is an implementation detail which, while useful when using the `authenticate` method, is not relevant in the `persistence.auth` package.

Maybe make it private inside the `FirebaseAuth` class ?

File: mobile/src/main/java/ch/epfl/sdp/mobile/infrastructure/persistence/auth/firebase/FirebaseAuth.kt [hunk 8]
```diff
@@ -38,28 +39,41 @@ class FirebaseAuth(private val actual: ActualFirebaseAuth) : Auth {
    */
   private suspend fun authenticate(
       block: suspend () -> AuthResult?,
+      mode: AuthenticationMode,
```
Comment (alexandrepiveteau) [inline mobile/src/main/java/ch/epfl/sdp/mobile/infrastructure/persistence/auth/firebase/FirebaseAuth.kt:42]: Keep the lambda parameter last so we can use some trailing lambda notation (like `authenticate(SignIn) { /* Something */ }`).

```suggestion
      mode: AuthenticationMode,
      block: suspend () -> AuthResult?,
```

**Review by alexandrepiveteau — CHANGES_REQUESTED** [review]
The most important change missing would be to remove this `AuthenticationMode` from the public API.

Additionally, you should probably bump up the coverage a bit on this PR.

File: mobile/src/main/java/ch/epfl/sdp/mobile/infrastructure/persistence/auth/Auth.kt (hunk 6 above)
Comment (alexandrepiveteau) [reply mobile/src/main/java/ch/epfl/sdp/mobile/infrastructure/persistence/auth/Auth.kt:40]: This makes checking for failure much easier :

```
val result: AuthenticationResult = // TODO ...
if (result is AuthenticationResult.Failure) {
  println("Something went wrong")
}

File: mobile/src/main/java/ch/epfl/sdp/mobile/infrastructure/persistence/auth/firebase/FirebaseAuth.kt (hunk 1 above)
Comment (Fouad-sys) [reply mobile/src/main/java/ch/epfl/sdp/mobile/infrastructure/persistence/auth/firebase/FirebaseAuth.kt]: Fixed in a362d0a

File: mobile/src/androidTest/java/ch/epfl/sdp/mobile/test/application/AuthenticationFacadeTest.kt (hunk 2 above)
Comment (Fouad-sys) [reply mobile/src/androidTest/java/ch/epfl/sdp/mobile/test/application/AuthenticationFacadeTest.kt]: Fixed in a362d0a

File: mobile/src/main/java/ch/epfl/sdp/mobile/infrastructure/persistence/auth/Auth.kt (hunk 6 above)
Comment (Fouad-sys) [reply mobile/src/main/java/ch/epfl/sdp/mobile/infrastructure/persistence/auth/Auth.kt]: Fixed in 12d715c

File: mobile/src/main/java/ch/epfl/sdp/mobile/infrastructure/persistence/auth/AuthenticationMode.kt (hunk 7 above)
Comment (Fouad-sys) [reply mobile/src/main/java/ch/epfl/sdp/mobile/infrastructure/persistence/auth/AuthenticationMode.kt]: Fully agree, fixed in 12d715c

File: mobile/src/main/java/ch/epfl/sdp/mobile/infrastructure/persistence/auth/firebase/FirebaseAuth.kt (hunk 8 above)
Comment (Fouad-sys) [reply mobile/src/main/java/ch/epfl/sdp/mobile/infrastructure/persistence/auth/firebase/FirebaseAuth.kt]: Fixed in 12d715c

File: mobile/src/main/java/ch/epfl/sdp/mobile/infrastructure/persistence/auth/firebase/FirebaseAuth.kt [hunk 9]
```diff
@@ -32,48 +31,59 @@ class FirebaseAuth(private val actual: ActualFirebaseAuth) : Auth {
...
+  }
+
   /**
    * Runs the [block] and maps the resulting [AuthResult] to an [AuthenticationResult].
    *
    * @param block the block of code which is used to interact with Firebase.
    */
   private suspend fun authenticate(
-      block: suspend () -> AuthResult?,
       mode: AuthenticationMode,
```
Comment (alexandrepiveteau) [inline mobile/src/main/java/ch/epfl/sdp/mobile/infrastructure/persistence/auth/firebase/FirebaseAuth.kt:50]: Missing some documentation for the `mode` parameter.

File: mobile/src/main/java/ch/epfl/sdp/mobile/ui/i18n/LocalizedStrings.kt [hunk 10]
```diff
@@ -20,6 +20,11 @@ interface LocalizedStrings {
   val authenticateSwitchToLogIn: String
   val authenticateSwitchToRegister: String
   val authenticateErrorFailure: String
+  val authenticationBadPasswordFailure: String
+  val authenticationWrongEmailFormatFailure: String
+  val authenticationInvalidUserFailure: String
+  val authenticationIncorrectPasswordFailure: String
+  val authenticationExistingAccountFailure: String
```
Comment (alexandrepiveteau) [inline mobile/src/main/java/ch/epfl/sdp/mobile/ui/i18n/LocalizedStrings.kt:27]: I think these strings should be prefixed with `authenticate` to remain consistent with the rest.

**Review by alexandrepiveteau — APPROVED** [review]

File: mobile/src/main/java/ch/epfl/sdp/mobile/infrastructure/persistence/auth/firebase/FirebaseAuth.kt [hunk 11]
```diff
@@ -31,35 +31,59 @@ class FirebaseAuth(private val actual: ActualFirebaseAuth) : Auth {
...
+        when (other) {
+          is FirebaseAuthWeakPasswordException -> AuthenticationResult.Failure.BadPassword
+          is FirebaseAuthInvalidCredentialsException ->
+              when (mode) {
+                AuthenticationMode.SignIn -> AuthenticationResult.Failure.IncorrectPassword
+                AuthenticationMode.SignUp -> AuthenticationResult.Failure.IncorrectEmailFormat
+              }
+          is FirebaseAuthUserCollisionException -> AuthenticationResult.Failure.ExistingAccount
+          is FirebaseAuthInvalidUserException -> AuthenticationResult.Failure.InvalidUser
+          else -> AuthenticationResult.Failure.Internal
```
Comment (KurohanaJuri) [inline mobile/src/main/java/ch/epfl/sdp/mobile/infrastructure/persistence/auth/firebase/FirebaseAuth.kt:68]: Nitpick : Import `AuthenticationResult.Failure.*` to have something more readable

**Review by KurohanaJuri — APPROVED** [review]
Some test names are a bit long, try next time to be more "compact"

**Pull request merged**